### PR TITLE
feat: handles cleanup by using owner ref and cleanup hook

### PR DESCRIPTION
**Pull request by bartoszmajsak**
### `OssmResourceTracker` 

A cluster-scoped resource for tracking objects created by OSSM installer. Primarily used as owner reference for resources created across namespaces so that they can be garbage collected by Kubernetes when they are not needed anymore.

### Cleanup funcs

OSSM Installer keeps a chain of functions that will be invoked during the cleanup phase. Right now it includes:

- deletion of `OssmResourceTracker` so the owned resources will be garbage collected on `KfDef` deletion
    - this included `ConfigMap`s, `EnvoyFilter`s and so on - essentially every ns-scoped resource created by the installer
- deletion of OAuth client. This is needed even though it also has `OwnerReferences` pointing to the tracker resource. OAuth client is also a cluster-scoped resource so it won't be garbage collected.
- removal of token volume from Istio ingress
- removal of extAuthzProvider

With this in place, subsequent deployments of `KfDef` wouldn't result in `OAuth flow failed` error which is caused by misaligned secrets between filter and client.

Those cleanup functions are invoked by a hook in Delete phase of `KfDef` operator.

### Open points

- we are not restoring everything yet. For example, settings such as IOR or auto-injection are not rolled back. We could use `OssmTrackerResource` (we might want to change its name btw ;)) to keep old values in a `.spec` and use it to restore on deletion.
- code need a bit more polish
- we need tests :)

### CODE REVIEW

File: apis/ossm.plugins.kubeflow.org/v1alpha1/ossm_types.go [hunk 1]
```diff
@@ -17,20 +17,22 @@ type OssmPlugin struct {
...
+// MeshSpec holds information on how Service Mesh should be configured.
 type MeshSpec struct {
 	Name        string   `json:"name,omitempty"`
 	Namespace   string   `json:"namespace,omitempty"`
 	Certificate CertSpec `json:"certificate,omitempty"`
 }
 
 type CertSpec struct {
-	Name     string `json:"name,omitempty" default:"opendatahub-self-signed-cert"`
+	Name     string `json:"name,omitempty"`
```
Comment (bartoszmajsak) [inline apis/ossm.plugins.kubeflow.org/v1alpha1/ossm_types.go:35]: We don't use defaults here but in the internal copy of the API/CRD.

File: main.go [hunk 2]
```diff
@@ -156,7 +156,7 @@ func main() {
 		os.Exit(1)
 	}
 
-	setupLog.Info("starting manager 123 !!!!")
+	setupLog.Info("starting manager")
```
Comment (bartoszmajsak) [inline main.go:159]: Poor man's debug ;)

File: pkg/kfapp/ossm/cert.go [hunk 3]
```diff
@@ -22,13 +21,22 @@ import (
...
 	}
 
+	objectMeta.SetOwnerReferences([]metav1.OwnerReference{
+		{
+			APIVersion: o.tracker.APIVersion,
+			Kind:       o.tracker.Kind,
+			Name:       o.tracker.Name,
+			UID:        o.tracker.UID,
+		},
+	})
```
Comment (bartoszmajsak) [inline pkg/kfapp/ossm/cert.go:38]: Note to self: we have it in multiple places - worth extracting

File: pkg/kfapp/ossm/k8s_resources.go
Comment (bartoszmajsak) [inline pkg/kfapp/ossm/k8s_resources.go:1]: Duplicated by purpose - we want to have our code self-contained and thus we should minimize coupling with other parts of the current operator as much as possible. Another benefit is that now we can refactor it to our liking, as I think we don't need all of the logic from the original `k8s_utils.go`

File: pkg/kfapp/ossm/cleanup.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+		},
+	}
+
+	c, err := dynamic.NewForConfig(o.config)
+	if err != nil {
+		return err
+	}
+
+	gvr := schema.GroupVersionResource{
+		Group:    "ossm.plugins.kubeflow.org",
```
Comment (cam-garrison) [inline pkg/kfapp/ossm/cleanup.go:51]: noting this Group as something to potentially change in future based on our call

**Review by cam-garrison — APPROVED** [review]
Nice :) excited to have cleanup working.

**Pull request merged**